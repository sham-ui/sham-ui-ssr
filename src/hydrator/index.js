import Storage from './storage/service';
import ServerSideHooks from './hooks/service';
import ServerSideDOM from './dom/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Main function for setup hydrating components
 * @param {Object} DI App DI container
 */
export function setup( DI ) {
    const storage = new Storage();

    // Override default hooks service
    DI.bind(
        HOOKS_DI_KEY,

        // Pass default hooks service for proxy some methods (rehydrate & resolveID)
        new ServerSideHooks( DI.resolve( HOOKS_DI_KEY ), storage )
    );

    // Override default dom service
    DI.bind(
        DOM_DI_KEY,

        // Don't pass default dom service, because ServerSideDOM full override it
        new ServerSideDOM( storage )
    );

    return storage.root;
}


/**
 * Hydrate all components
 * @param {Object} DI App DI container
 * @return {Promise}
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import { setupHydrator, hydrate } from 'sham-ui-ssr/hydrator';
 *
 * const DI = createDI();
 * const root = setupHydrator( DI );
 * mainInitializer( DI, root );
 * start( DI );
 * hydrate( DI ).then( storage => {
 *     const { html, data } = storage.hydrate();
 * } )
 */
export function hydrate( DI ) {
    return DI.resolve( HOOKS_DI_KEY ).hydrating();
}
