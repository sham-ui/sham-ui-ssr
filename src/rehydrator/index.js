import RestoreDom from './dom/service';
import RehydrateHooks from './hooks/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Save hooks & dom services for restore after rehydrating
 * @param {Object} DI App DI container
 * @return {Function}
 */
function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

/**
 * Main function for rehydrating
 * @param {Object} DI App DI container
 * @param {Object} storage
 * @return {Function} Restore function for disable rehydrating and switch to default behavior
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import setupUnsafe from 'sham-ui-unsafe';
 * import { setupRehydrator } from 'sham-ui-ssr/rehydrator';
 *
 * const DI = createDI();
 * const data = JSON.parse( document.getElementById( 'data' ).innerText );
 * const disableRehydrating = setupRehydrator( DI, data );
 * mainInitializer( DI, root );
 * start( DI );
 * disableRehydrating();
 * setupUnsafe( DI );
 */
export function setup( DI, storage ) {
    const restore = makeRestore( DI );
    DI.bind( HOOKS_DI_KEY, new RehydrateHooks( storage ) );
    DI.bind( DOM_DI_KEY, new RestoreDom( storage ) );
    return restore;
}
