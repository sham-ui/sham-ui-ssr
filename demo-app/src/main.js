import './styles/main.scss';
import setupUnsafe from 'sham-ui-unsafe';
import { createDI, start } from 'sham-ui';
import { setup as setupHydrator, hydrate } from '../../src/hydrator/index';
import { setup as setupRehydrator } from '../../src/rehydrator/index';
import mainInitializer from './initializers/main';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

const DI = createDI();
const restoreOriginal = makeRestore( DI );

const root = setupHydrator( DI );
mainInitializer( DI, root );
start( DI );
hydrate( DI ).then(
    ( storage ) => {
        const getById = document.getElementById.bind( document );
        // eslint-disable-next-line no-console
        console.dir( storage );
        try {
            const { html, data } = storage.hydrate();
            getById( 'ssr-rendered' ).innerHTML = html;
            getById( 'raw-html' ).innerText = html;
            getById( 'data' ).innerText = JSON.stringify( JSON.parse( data ), null, 4 );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
        try {
            DI.resolve( 'sham-ui:store' ).byId.clear();
            const root = getById( 'client' );
            root.innerHTML = getById( 'ssr-rendered' ).innerHTML;
            restoreOriginal();
            const data = JSON.parse( document.getElementById( 'data' ).innerText );
            const disableRehydrating = setupRehydrator( DI, data );
            mainInitializer( DI, root );
            start( DI );
            disableRehydrating();
            setupUnsafe( DI );
            // eslint-disable-next-line no-console
            console.dir( DI.resolve( 'sham-ui:store' ).byId.keys() );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
    }
);

