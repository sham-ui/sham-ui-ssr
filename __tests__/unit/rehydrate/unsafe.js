import { compile } from 'sham-ui-test-helpers';
import { ssrAndRehydrate } from '../../../src/testing';

it( 'should insert constants as HTML', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`
            <div>
                {% unsafe "<br>" %}
            </div>
        `
    );
    expect( meta.html ).toBe( '<div><br></div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should insert variables as HTML', async() => {
    expect.assertions( 3 );
    const meta = await ssrAndRehydrate(
        compile`
            <div>
                {% unsafe html %}
            </div>
        `
    );
    expect( meta.html ).toBe( '<div></div>' );

    meta.component.update( { html: '<a href="javascript:XSS;">Link</a>' } );
    expect( meta.ctx.container.innerHTML ).toBe(
        '<div><a href="javascript:XSS;">Link</a></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should remove old DOM nodes and insert new', async() => {
    expect.assertions( 5 );
    const meta = await ssrAndRehydrate(
        compile`
            <div>
                {% unsafe html %}
            </div>
        `,
        {
            html: '<div>foo</div><br>'
        }
    );
    expect( meta.html ).toBe( '<div><div>foo</div><br></div>' );

    meta.component.update( { html: '<input type="datetime"><hr><div>baz</div>' } );
    expect( meta.ctx.container.innerHTML ).toBe(
        '<div><input type="datetime"><hr><div>baz</div></div>'
    );

    meta.component.update( { html: '' } );
    expect( meta.ctx.container.innerHTML ).toBe( '<div></div>' );

    meta.component.update( { html: '<!-- comment -->' } );
    expect( meta.ctx.container.innerHTML ).toBe( '<div><!-- comment --></div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should insert unsafe with placeholders', async() => {
    expect.assertions( 3 );
    const meta = await ssrAndRehydrate(
        compile`
            <div>
                {% unsafe "<br>" %}
            </div>
            <div>
                {% unsafe html %}
            </div>
        `,
        {
            html: '<hr>'
        }
    );
    expect( meta.html ).toBe( '<div><br></div><div><hr></div>' );

    meta.component.update( { html: '<br><!-- comment --><link href="http://ShamUIView.js.org">' } );
    expect( meta.ctx.container.innerHTML ).toBe(

        // eslint-disable-next-line max-len
        '<div><br></div><div><br><!-- comment --><link href="http://ShamUIView.js.org"></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'if with unsafe tag', async() => {
    expect.assertions( 4 );
    const meta = await ssrAndRehydrate(
        compile`
            <div>
                {% if test %}
                    <div>
                        {% unsafe "<i>unsafe</i>" %}
                    </div>
                {% endif %}
            </div>
        `,
        {
            test: true
        }
    );
    expect( meta.html ).toBe( '<div><div><i>unsafe</i></div></div>' );

    meta.component.update( { test: false } );
    expect( meta.ctx.container.innerHTML ).toBe( '<div></div>' );

    meta.component.update( { test: true } );
    expect( meta.ctx.container.innerHTML ).toBe( '<div><div><i>unsafe</i></div></div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work with first level non-elements', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`
            text
            {% if cond1 %}
                <div class="if">ok</div>
            {% endif %}
            {% for loop1 %}
                <div class="for">ok</div>
            {% endfor %}
            <div on="{{ tag }}">
                <div class="custom">ok</div>
            </div>
            <div>
                {% unsafe "<i class='unsafe'>" + xss + "</i>" %}
            </div>
        `,
        {
            cond1: true,
            loop1: [ 1, 2, 3 ],
            tag: true,
            xss: 'ok'
        }
    );
    expect( meta.html ).toBe(
        //eslint-disable-next-line max-len
        'text <div class="if">ok</div><!--0--><div class="for">ok</div><div class="for">ok</div><div class="for">ok</div><!--1--><div on="true"><div class="custom">ok</div></div><div><i class=\'unsafe\'>ok</i></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );
