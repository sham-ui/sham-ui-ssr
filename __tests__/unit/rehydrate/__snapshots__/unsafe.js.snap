// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if with unsafe tag 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "test": true,
  },
  "html": "
<div>
  <div>
    <i>unsafe</i>
  </div>
</div>
",
}
`;

exports[`should insert constants as HTML 1`] = `
Object {
  "Options": Object {
    "ID": "component",
  },
  "html": "
<div>
  <br>
</div>
",
}
`;

exports[`should insert unsafe with placeholders 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "html": "<br><!-- comment --><link href=\\"http://ShamUIView.js.org\\">",
  },
  "html": "
<div>
  <br>
</div>
<div>
  <br><!-- comment -->
  <link href=\\"http://ShamUIView.js.org\\">
</div>
",
}
`;

exports[`should insert variables as HTML 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "html": "<a href=\\"javascript:XSS;\\">Link</a>",
  },
  "html": "
<div>
  <a href=\\"javascript:XSS;\\">Link</a>
</div>
",
}
`;

exports[`should remove old DOM nodes and insert new 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "html": "<!-- comment -->",
  },
  "html": "
<div>
  <!-- comment -->
</div>
",
}
`;

exports[`should work with first level non-elements 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "cond1": true,
    "loop1": Array [
      1,
      2,
      3,
    ],
    "tag": true,
    "xss": "ok",
  },
  "html": "
text <div class=\\"if\\">ok</div>
<!--0-->
<div class=\\"for\\">ok</div>
<div class=\\"for\\">ok</div>
<div class=\\"for\\">ok</div>
<!--1-->
<div on=\\"true\\">
  <div class=\\"custom\\">ok</div>
</div>
<div>
  <i class=\\"unsafe\\">ok</i>
</div>
",
}
`;
